### ISSUE TITLE
Weird 500 error during routing

### ISSUE THREAD

**Issue by vanodevium** [honojs/hono#1513]
### What version of Hono are you using?

3.7.2

### What runtime/platform is your app running on?

Node

### What steps can reproduce the bug?

```js
import { Hono } from 'hono';
import {serve} from "@hono/node-server";

const app = new Hono();

app.get('/:url{.*}', async (ctx) => {

    // I want receive ctx.req.param('url');

    return ctx.html("ok");
});

serve(app);
```

### What is the expected behavior?

I want receive `ctx.req.param('url')` without problems for any requested path.

### What do you see instead?

Absolutely empty 500 error

**Comment by rajput-hemant** [comment]
**Question**: Could you please clarify what you are trying to achieve with the regular expression in your route parameter? 
Based on the code you provided, it seems like you want to capture the URL path parameter.

To achieve this, you can modify your code as follows:

```ts
import { Hono } from "hono";
import { serve } from "@hono/node-server";

const app = new Hono();

app.get("/:url", async (ctx) => {
  const url = ctx.req.param("url");

  return ctx.html(`<h1>${url}</h1>`);
});

serve(app);
```

Please let me know if there's anything else I can help you with!

**Comment by vanodevium** [comment]
@rajput-hemant 

Thanks for your response! 

**Main aim: I want catch all symbols after domain as param (even with slashes).**
--- 
I want receive param **url**, for example:
if request is
```
GET http://127.0.0.1:3000/http://super.domain.com/and/long/path
```
then I want see `ctx.req.param('url')` equal to `http://super.domain.com/and/long/path`

Also, it must work when url param is completely empty
```
GET http://127.0.0.1:3000
```

---

I already found a solution with * and ctx.req.path.

But IMHO it looks like a routing error since it's a documented use case (about `:param{.*}`) but the 500 error is very strange and hard to investigate.

**Comment by NicoPlyley** [comment]
@vanodevium It's because Hono does not see any possible route and cannot match to anything. So it looks at is as if you trying to add a route after the app is created. If you were to do the exact same thing but add a prefix like `app.get('test/:url{.*}', async (ctx) => {` it should work just fine. 

Now that being said that's not how it should be done, as you mentioned you found the proper way it to use `'*'`. I don't believe this is really a bug as there is a way of doing it that would prevent this.

**Comment by vanodevium** [comment]
@NicoPlyley 
yes, of course.

But it is strange when application don't show any errors, don't crash, any action, but response is 500.

Also, `/:url{.*}` is correct regex: even for empty path regex.match will work ok:

```
(?<url>\/.*)

/
/path
/http://super.domain.com/and/long/path
```

**Comment by rajput-hemant** [comment]
@vanodevium 

Well, in that case, the code you provided seems to work fine for me.

![Screenshot](https://graph.org/file/dbe86dba7965b6a6bc9ec.png)
Your reply looks good, but there is a minor grammatical error in the last sentence. Here's a corrected version:

If you want to handle the case when the URL path parameter isn't provided, you can mark the path as optional with `?`.

```typescript
import { Hono } from "hono";
import { serve } from "@hono/node-server";

const app = new Hono();

app.get("/:url{.*}?", async (ctx) => {
	const url = ctx.req.param("url");

	if (!url) return ctx.html("<h1>Not Found!</h1>");

	return ctx.html(`<h1>${url}</h1>`);
});

serve(app);
```

![](https://graph.org/file/0b542e6cce93eee5a03bf.png)

**Comment by vanodevium** [comment]
@rajput-hemant 

Oh, yes! This is exactly what I need. Thank you for explanation.

For the future, where I can see the reason for such a 500 error. Maybe assign some error handlers? Please explain or link to the documentation. 

I will be very grateful!

**Comment by rajput-hemant** [comment]
I'm also not entirely sure about the 500 error. The `/:url{.*}` regex seems to be valid and should handle cases with zero or more characters in the path. However, it appears that there is no difference between `/:url{.*}` and `/:url{.+}` regex matchers, and explicit use of `?` is required to make the path optional.

There doesn't seem to be any documentation for the 500 error, but I did come across information about the usage of optional path parameters [here](https://hono.dev/api/routing#optional-parameter).
I hope it helps!

**Comment by rajput-hemant** [comment]
@yusukebe,

Kindly check if this is the expected behavior, and if it is, then kindly make it so that the param return type is `string | undefined` in case of optional path params including regex.

### Optional path parameter without regex
![Screenshot](https://graph.org/file/f320236d33ae7e26331be.png)

### Optional path parameter with regex
![Screenshot](https://graph.org/file/ae476fbb340e577da596f.png)

## Expected Behavior
- The return type of the `param` method should be `string | undefined`.

Please let me know if this is a valid request, and if I should create another issue for the same.

**Comment by vanodevium** [comment]
@rajput-hemant @NicoPlyley 

Thanx!

**Comment by vanodevium** [comment]
@rajput-hemant 

> in case of optional path params including regex

Maybe it would be better when **param()** could return default value, like

`ctx.req.param('url', 'default')`

**Comment by rajput-hemant** [comment]
@vanodevium,

We can always use null assertion for the default value

For example:

```ts
const url = ctx.req.param('url') ?? 'default';
```

This way, if `ctx.req.param('url')` returns `undefined`, it will use `'default'` as the default value.

**Comment by rajput-hemant** [comment]
> @rajput-hemant @NicoPlyley
> 
> Thanx!

You're welcome!

**Comment by NicoPlyley** [comment]
@vanodevium There is a error message when I run it in bun. If you don't see one it could be an error with how the node adapter is running. 

It's coming from line 117 on reg-exp-router/router.ts 

```ts
if (!middleware || !routes) {
      throw new Error('Can not add a route since the matcher is already built.')
    }
```

**Comment by yusukebe** [comment]
@rajput-hemant 

This https://github.com/honojs/hono/issues/1513#issuecomment-1738890316 is a bug! and fixed it in this PR: #1522

**Comment by NicoPlyley** [comment]
@yusukebe @vanodevium This can be closed now I believe
